### PR TITLE
Doctor notes enhancement

### PR DESCRIPTION
**Pull request by UdaySagar-Git**
## Proposed Changes

- Adds mention support 
- Adds fields to track root_note and child notes for thread view 
- Adds notification support upon mentioning user

### Associated Issue

- Link to issue here, explain how the proposed solution will solve the reported issue/ feature request.

### Architecture changes

- Remove this section if not used

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@ohcnetwork/care-backend-maintainers @ohcnetwork/care-backend-admins

## Release Notes

- **New Features**
  - Enhanced handling of patient notes, including support for replies, mentions of users, and hierarchical linking of notes.
  - Introduced a notification system for mentions in patient notes, improving user engagement.
  - Added a new file type for patient notes in file uploads.
  - New fields in patient note serializers to improve detail and context.
  - Improved archival management for file uploads.
  - New functionality to track the last edited information for patient notes.
  - Users can now associate files with patient notes, enhancing interactivity.

- **Bug Fixes**
  - Implemented stricter access control for editing patient notes to ensure only the creator can modify them.

- **Documentation**
  - Updated expected structure for patient note API responses to include new fields.

These updates significantly improve the user experience by enhancing note-taking capabilities and notification systems.

### CODE REVIEW

**Comment by UdaySagar-Git** [comment]
@sainak can you review this pr

File: care/facility/models/notification.py [hunk 1]
```diff
@@ -41,6 +41,7 @@ class Event(enum.Enum):
         PUSH_MESSAGE = 220
         PATIENT_PRESCRIPTION_CREATED = 230
         PATIENT_PRESCRIPTION_UPDATED = 240
+        PATIENT_NOTE_MENTIONED = 250
```
Comment (rithviknishad) [inline care/facility/models/notification.py]: nitpick, just more readable...

```suggestion
        MENTIONED_IN_PATIENT_NOTE = 250
```

Comment (UdaySagar-Git) [reply care/facility/models/notification.py]: done

File: care/facility/models/patient.py [hunk 2]
```diff
@@ -806,13 +807,26 @@ class PatientNotes(FacilityBaseModel, ConsultationRelatedPermissionMixin):
...
         # and hence the permission mixin will fail if edit/object_read permissions are checked (although not used as of now)
         # Remove once patient notes is made consultation specific.
         return self
 
+    @property
+    def mentioned_users(self):
+        # handling both - and _ (valid usernames : devdoctor, dev-doctor and dev_doctor)
+        usernames = set(re.findall(r"@([a-zA-Z0-9_-]+)", self.note))
+        return User.objects.filter(username__in=usernames)
+
```
[the remark this answers is not in the document]
Comment (UdaySagar-Git) [reply care/facility/models/patient.py]: resolved

File: care/utils/notification_handler.py [hunk 3]
```diff
@@ -300,29 +306,41 @@ def generate_cause_objects(self):
...
+        mentioned_users = self.mentioned_users
+
+        if self.generate_for_facility:
+            facility_users = FacilityUser.objects.filter(facility_id=self.facility.id)
+            if self.event != Notification.Event.MESSAGE:
+                facility_users.exclude(
+                    user__user_type__in=(
+                        User.TYPE_VALUE_MAP["Staff"],
+                        User.TYPE_VALUE_MAP["StaffReadOnly"],
+                    )
```
[the remark this answers is not in the document]
Comment (UdaySagar-Git) [reply care/utils/notification_handler.py]: updated

File: care/facility/migrations/0468_patientnotes_root_note_alter_fileupload_file_type_and_more.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    dependencies = [
+        ('facility', '0467_alter_hospitaldoctors_area'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='patientnotes',
+            name='root_note',
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='child_notes', to='facility.patientnotes'),
+        ),
```
[the remark this answers is not in the document]
Comment (UdaySagar-Git) [reply care/facility/migrations/0468_patientnotes_root_note_alter_fileupload_file_type_and_more.py:18]: added checks for circular references

File: care/facility/tests/test_patient_api.py [hunk 5]
```diff
@@ -327,6 +332,197 @@ def test_patient_note_edit(self):
...
+        response = self.client.get(
+            f"/api/v1/patient/{self.patient.external_id}/notes/{note_id}/"
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data["files"]), len(test_files))
+        file_names = [f["name"] for f in response.data["files"]]
+        for file_data in test_files:
+            self.assertIn(file_data["name"], file_names)
+
```
[the remark this answers is not in the document]
Comment (UdaySagar-Git) [reply care/facility/tests/test_patient_api.py:404]: this is a part of file upload api

File: care/facility/api/serializers/patient.py [hunk 6]
```diff
@@ -454,9 +455,35 @@ class Meta:
...
+            ),
+            many=True,
+        ).data
+    return None
+
+
 class ReplyToPatientNoteSerializer(serializers.ModelSerializer):
     id = serializers.CharField(source="external_id", read_only=True)
     created_by_object = UserBaseMinimumSerializer(source="created_by", read_only=True)
+    reply_to = serializers.CharField(source="reply_to.external_id", read_only=True)
```
Comment (rithviknishad) [inline care/facility/api/serializers/patient.py]:
```suggestion
    reply_to = serializers.UUIDField(source="reply_to.external_id", read_only=True)
```

File: care/facility/models/patient.py [hunk 7]
```diff
@@ -806,13 +807,28 @@ class PatientNotes(FacilityBaseModel, ConsultationRelatedPermissionMixin):
...
         # Remove once patient notes is made consultation specific.
         return self
 
+    @property
+    def mentioned_users(self):
+        if not hasattr(self, "_mentioned_users_cache"):
+            # handling both - and _ (valid usernames : devdoctor, dev-doctor and dev_doctor)
+            usernames = set(re.findall(r"@([a-zA-Z0-9][-_a-zA-Z0-9]{0,38})", self.note))
+            self._mentioned_users_cache = User.objects.filter(username__in=usernames)
+        return self._mentioned_users_cache
```
Comment (rithviknishad) [inline care/facility/models/patient.py]: thoughts on using django's caching framework here instead?

the cache here currently used would only be instance specific right?

cc: @sainak @vigneshhari

File: care/facility/api/serializers/patient.py [hunk 8]
```diff
@@ -491,12 +521,52 @@ class PatientNotesSerializer(serializers.ModelSerializer):
         allow_null=True,
     )
     reply_to_object = ReplyToPatientNoteSerializer(source="reply_to", read_only=True)
+    files = serializers.SerializerMethodField()
+    replies = ReplyToPatientNoteSerializer(many=True, read_only=True)
+    child_notes = ReplyToPatientNoteSerializer(many=True, read_only=True)
+    root_note_object = ReplyToPatientNoteSerializer(source="root_note", read_only=True)
+    mentioned_users = UserBaseMinimumSerializer(many=True, read_only=True)
```
[the remark this answers is not in the document]
Comment (UdaySagar-Git) [reply care/facility/api/serializers/patient.py:528]: done

File: care/facility/models/patient.py (hunk 7 above)
Comment (UdaySagar-Git) [reply care/facility/models/patient.py]: done, switched to django cache

File: care/facility/models/patient.py [hunk 9]
```diff
@@ -806,13 +808,34 @@ class PatientNotes(FacilityBaseModel, ConsultationRelatedPermissionMixin):
...
+
+        if not hit:
+            # handling both - and _ (valid usernames : devdoctor, dev-doctor and dev_doctor)
+            usernames = set(re.findall(r"@([a-zA-Z0-9][-_a-zA-Z0-9]{0,38})", self.note))
+            users = User.objects.filter(username__in=usernames)
+            cache.set(key, users, 60 * 60)  # 1 hour
+            return users
+
+        return hit
+
```
[the remark this answers is not in the document]
Comment (UdaySagar-Git) [reply care/facility/models/patient.py:838]: shouldn't be an issue since it gets invalidated in 1 hr?

[the remark this answers is not in the document]
Comment (UdaySagar-Git) [reply care/facility/models/patient.py]: done made changes for cachekey to include modified date

**Review by rithviknishad — APPROVED** [review]
lgtm

**Pull request merged**